src/pages/studyView/StudyViewPageStore.ts

```diff
@@ -501,26 +537,129 @@ export class StudyViewPageStore {
                     this.studyIds.slice()
                 ); // slice because cant pass mobx
+
                 return groups.map(group =>
                     Object.assign(
                         group.data,
-                        { uid: group.id },
+                        { uid: group.id, isSharedGroup: false },
                         finalizeStudiesAttr(group.data, this.sampleSet.result!)
                     )
                 );
-            } else {
-                return [];
             }
+            return [];
+        },
+        default: [],
+    });
+
+    readonly sharedGroups = remoteData<StudyViewComparisonGroup[]>({
+        await: () => [this.sampleSet, this.queriedPhysicalStudyIds],
+        invoke: async () => {
+            const promises: Promise<Group>[] = [];
+            Object.keys(this.sharedGroupSet).forEach(groupId => {
+                promises.push(comparisonClient.getGroup(groupId));
+            });
+            const studyIdsSet = stringListToSet(
+                this.queriedPhysicalStudyIds.result!
+            );
+            const groups = await Promise.all(promises);
+            return groups
+                .filter(
+                    group =>
+                        !_.some(
+                            group.data.studies,
+                            study => studyIdsSet[study.id] === undefined
+                        )
+                )
+                .map(group =>
+                    Object.assign(
+                        group.data,
+                        { uid: group.id, isSharedGroup: true },
+                        finalizeStudiesAttr(group.data, this.sampleSet.result!)
+                    )
+                );
+        },
+        default: [],
+    });
+
+    readonly comparisonGroups = remoteData<StudyViewComparisonGroup[]>({
+        await: () => [this.userSavedGroups, this.sharedGroups],
+        invoke: async () => {
+            let groups: StudyViewComparisonGroup[] = _.cloneDeep(
+                this.userSavedGroups.result
+            );
+            let groupIdSet: { [s: string]: boolean } = {};
+
+            const groupsToSave: string[] = [];
+            if (this.sharedGroups.result.length > 0) {
+                groupIdSet = stringListToSet(groups.map(group => group.uid));
+                this.sharedGroups.result.forEach(sharedGroup => {
+                    if (groupIdSet[sharedGroup.uid] === undefined) {
+                        groups.push(sharedGroup);
+                        groupIdSet[sharedGroup.uid] = true;
+                        if (this.isLoggedIn) {
+                            groupsToSave.push(sharedGroup.uid);
+                        }
+                    }
+                });
+            }
+
+            if (groupsToSave.length > 0) {
+                const addPromises: Promise<void>[] = [];
+                groupsToSave.forEach(groupId =>
+                    addPromises.push(comparisonClient.addGroupToUser(groupId))
```

Comment:
I don't think `invoke` should be executing changes like this. unless it's necessary for the logic of `invoke` it should be in `onResult`

Reply:
agreed.    i think onresult is the way to go.   could we reuse logic from the login reaction?

Reply:
yeah, if that reaction just referenced comparisonGroups, would that not cover all the cases?

Reply:
I tried referencing `comparisonGroups` in reaction but it wasn't working as expected. Anyways I ended removing the reaction as it is no more required. And I ended up using `onResult` which works fine for all cases